### PR TITLE
Fix Socket.ReceiveFrom on Unix

### LINKED ISSUE #45651: Socket ReceiveFrom() and SendTo() issues on Linux after .NET 5 upgrade
### Description
An open source DNS server project that I am working currently uses .NET 3.1 and works well on Linux and Windows. I upgraded the project to .NET 5 and it stop functioning immediately without any kind of errors being generated. So I started investigating the issue and used Console.WriteLine and tcpdump on capture packets on all networks on the Ubuntu 20.04 VM.

### Configuration

* The code is running .NET 5. Earlier the same code worked perfectly on .NET 3.1.
* Tested on Ubuntu 20.04 LTS x64.
* OS is running in a VM with 1 CPU and 1GB RAM.

### Regression?

From the investigations, it seems there is an issue with how Socket handles ReceiveFrom() and SendTo() in .NET 5 upgrade. The ReceiveFrom() is giving wrong `remoteEP` values after a couple of packets. The remoteEP values keep on repeating in the Console.WriteLine() which meant that the method is giving wrong EndPoint for the received packet.

Further more, the SendTo() method sends the packets to a `remoteEP` different from the one provided in the argument. It too seems to gets stuck on the same end point after sending a couple of packets and then keeps on sending packets to destination that was not specified in the `remoteEP` argument.

This issue comes only on Linux and Windows seems to be not affected.

### Other information

Attaching PCAP files to inspect in a zip file: [dns-issue-pcap.zip](https://github.com/dotnet/runtime/files/5648857/dns-issue-pcap.zip)

1. `dns-issue1.pcap` : In here ICMP port unreachable can be observed as the DNS server responds to a request with the stuck end point "127.0.0.1:36145". Expected destination end point is "192.168.180.1:54290" for the last response in the pcap.

2. `dns-issue2.pcap`: In here the DNS server is sending response to a stuck end point where the IP seems to be correct but the port number is different ie. "192.168.180.1:58002". Expected destination end point is "192.168.180.1:62807" for the last response in the pcap.

The link to the affected DNS server project code to show the .NET 3.1 working implementation that broke with .NET 5 upgrade: https://github.com/TechnitiumSoftware/DnsServer/blob/master/DnsServerCore/Dns/DnsServer.cs#L224-L354

### PR DESCRIPTION
**Pull request by antonfirsov**
Fix #45651. The bug happens when a synchronous `ReceiveFrom` is being used after asynchronous `SendToAsync` call(s) on the same socket. As @tmds pointed out in https://github.com/dotnet/runtime/issues/45651#issuecomment-744637362, we did not set `SetReceiveFlags = true` in `SocketAsyncContext.ReceiveFrom`
https://github.com/dotnet/runtime/blob/31062e4ec6b248b709095e13947400b49d507e53/src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs#L1510-L1516

which resulted in `BufferedMemoryReceieveOperation` completion attempt with `recv` instead of `recvmsg`:
https://github.com/dotnet/runtime/blob/31062e4ec6b248b709095e13947400b49d507e53/src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs#L473-L479

Our existing UDP tests (`SendToRecvFrom_Datagram_UDP`) did not exercise `BufferMemoryReceiveOperation`, because completion of the synchronous `ReceiveFrom` always happened without the use of `_receiveQueue`:

https://github.com/dotnet/runtime/blob/31062e4ec6b248b709095e13947400b49d507e53/src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs#L1505-L1508

For reasons I don't understand, the `BufferMemoryReceiveOperation` path gets only executed if a synchronous `ReceiveFrom` is being called after an asynchronous operation. The PR adds tests for that case for multiple async variants.
Another weird thing: while my standalone repro resulted in in an assertion failure (https://github.com/dotnet/runtime/issues/45651#issuecomment-744631080), the new tests timeout (hang?) instead. (Both issues are fixed with `SetReceiveFlags = true`.)

We should probably consider adding similar tests for TCP send/receive.

### CODE REVIEW

**Comment by antonfirsov** [comment]
@tmds @geoffkizer please take a look. Note: starting from tomorrow, I will be on vacation until the 4th of January, looking at GitHub occasionally.

**Review by wfurt — APPROVED** [review]
LGTM

**Comment by geoffkizer** [comment]
> For reasons I don't understand, the BufferMemoryReceiveOperation path gets only executed if a synchronous ReceiveFrom is being called after an asynchronous operation. The PR adds tests for that case for multiple async variants.

This is the "sync-over-async" path in SocketAsyncContext.

What happens is this: The OS socket starts out in blocking mode. If you only perform sync IO, it will stay in blocking mode. But as soon as you perform an async IO, we need to switch to nonblocking mode at the OS level. But we still support performing sync operations, even though the underlying OS socket is nonblocking. We do this by simulating sync operations on top of the async infrastructure in SocketAsyncContext. So the sync caller will block on a ManualResetEventSlim that we signal when the operation can proceed, and from the user's point of view this looks and behaves like sync IO, even though under the covers it actually is done using a nonblocking socket and epoll.

We do have a bunch of tests for this. See `SocketHelperSyncForceNonBlocking`. Unfortunately, it seems that these tests only use the sync overloads that take byte[], not the ones that take Memory. We should probably add another case here.

**Review by geoffkizer — CHANGES_REQUESTED** [review]
The fix itself looks fine. (Edited:) But can we fix the tests as suggested above instead of adding a standalone test?

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs [hunk 1]
```diff
@@ -1511,6 +1511,7 @@ public SocketError ReceiveFrom(Memory<byte> buffer, ref SocketFlags flags, byte[
             {
                 Buffer = buffer,
                 Flags = flags,
+                SetReceivedFlags = true,
```
Comment (stephentoub) [inline src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs:1514]: Product fix LGTM

**Comment by antonfirsov** [comment]
@geoffkizer `ForceNonBlocking` seems to be sufficient to reproduce the issue. I pushed a change that applies it with the help of `SocketTestHelper`, PTAL.

**Comment by geoffkizer** [comment]
Are there other UDP tests that should be calling ConfigureNonBlocking? Seems like there must be.

**Comment by antonfirsov** [comment]
I was able to find only one more among the generic tests utilizing SocketTestHelper infrastructure (`UdpReceiveGetsCanceledByDispose`), pushed a change for that.

**Review by geoffkizer — APPROVED** [review]

**Comment by antonfirsov** [comment]
I don't know what "runtime-staging" is but it seems to fail because of an iOS build timing out, which is unrelated to this change.

**Pull request merged**